bin/run.py

```diff
@@ -41,34 +48,92 @@ def run(self, bar, *, interaction=None, submission_args=None):
                 self, interaction=interaction, submission_args=submission_args
             )
         else:
-            result = self.submission.run(self.testcase.in_path, self.out_path)
-            if result.duration > self.problem.settings.timelimit:
-                result.verdict = Verdict.TIME_LIMIT_EXCEEDED
-            elif result.status == ExecStatus.ERROR:
-                result.verdict = Verdict.RUNTIME_ERROR
-                if config.args.error:
-                    result.err = 'Exited with code ' + str(result.returncode) + ':\n' + result.err
-                else:
-                    result.err = 'Exited with code ' + str(result.returncode)
-            else:
-                # Overwrite the result with validator returncode and stdout/stderr, but keep the original duration.
-                duration = result.duration
-                result = self._validate_output(bar)
-                if result is None:
-                    error(f'No output validators found for testcase {self.testcase.name}')
-                    result = ExecResult(None, ExecStatus.REJECTED, 0, False, None, None)
-                    result.verdict = Verdict.VALIDATOR_CRASH
-                else:
-                    result.duration = duration
-
-                    if result.status:
+            if interaction:
+                assert not interaction.is_relative_to(self.tmpdir)
+                interaction = interaction.open('a')
+            nextpass = self.feedbackdir / 'nextpass.in' if self.problem.multipass else False
+            last_pass = 0
+            max_duration = 0
+            tle_result = None
+            while True:
+                last_pass += 1
+                result = self.submission.run(self.in_path, self.out_path)
+                max_duration = max(max_duration, result.duration)
+                if result.duration > self.problem.settings.timelimit:
+                    result.verdict = Verdict.TIME_LIMIT_EXCEEDED
+                    if tle_result is None:
+                        tle_result = result
+                        tle_result.pass_id = last_pass if self.problem.multipass else None
+                    else:
+                        tle_result.timeout_expired |= result.timeout_expired
+                elif result.status == ExecStatus.ERROR:
+                    result.verdict = Verdict.RUNTIME_ERROR
+                    if config.args.error:
+                        result.err = (
+                            'Exited with code ' + str(result.returncode) + ':\n' + result.err
+                        )
+                    else:
+                        result.err = 'Exited with code ' + str(result.returncode)
+
+                if interaction:
```

Comment:
add a comment here what this is doing, and when/how it is triggered.

Also, can you move this right below `result = self.submission.run(..)`?

Then the if-statement on line 91 can be simplified to just early-break in the `RTE` case above, and a conditional early break in the `TLE` case above.